ocaml/message-switch/core/make.ml

```diff
@@ -352,26 +352,36 @@ functor
 
     let listen ~process ~switch:port ~queue:name () =
       let token = Printf.sprintf "%d" (Unix.getpid ()) in
-      M.connect port >>= fun c ->
-      Connection.rpc c (In.Login token) >>|= fun (_ : string) ->
-      Connection.rpc c (In.CreatePersistent name) >>|= fun _ ->
+      let reconnect () =
+        M.connect port >>= fun request_conn ->
+        Connection.rpc request_conn (In.Login token) >>|= fun (_ : string) ->
+        Connection.rpc request_conn (In.CreatePersistent name) >>|= fun _ ->
+        M.connect port >>= fun reply_conn ->
+        Connection.rpc reply_conn (In.Login token) >>|= fun (_ : string) ->
+        Connection.rpc reply_conn (In.CreatePersistent name) >>|= fun _ ->
+        return (Ok (request_conn, reply_conn))
+      in
+      reconnect () >>|= fun ((request_conn, reply_conn) as c) ->
       let request_shutdown = M.Ivar.create () in
       let on_shutdown = M.Ivar.create () in
+      let mutex = M.Mutex.create () in
       let t = {request_shutdown; on_shutdown} in
       let rec loop c from =
         let transfer = {In.from; timeout; queues= [name]} in
         let frame = In.Transfer transfer in
-        let message = Connection.rpc c frame in
+        let message = Connection.rpc request_conn frame in
         any [map (fun _ -> ()) message; M.Ivar.read request_shutdown]
         >>= fun () ->
         if is_determined (M.Ivar.read request_shutdown) then (
           M.Ivar.fill on_shutdown () ; return (Ok ())
         ) else
           message >>= function
           | Error _e ->
-              M.connect port >>= fun c ->
-              Connection.rpc c (In.Login token) >>|= fun (_ : string) ->
-              loop c from
+              M.Mutex.with_lock mutex (fun () ->
```

Comment:
I think there is some duplication now with the mutex in Protocol_unix.ml
Would be useful to delete the code from Protocol_unix now that the mutex has been added to the shared code.

Reply:
I feel it is non-trivial to funtorise the `protocol_unix` module, maybe we could do that as a separate PR?